compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt

```diff
@@ -0,0 +1,424 @@
+        onDragStart: (Offset) -> Unit,
+        onDrag: (Offset) -> Unit,
+        onDragCancel: () -> Unit,
+        onDrop: (DropData) -> Unit
+    ): Int {
+        isActive = true
+
+        handlers[idsCounter] = ComponentDragHandler(componentCoordinates, onDragStart, onDrag, onDragCancel, onDrop)
+
+        if (isExternalDragInsideComponent(componentCoordinates, windowDragCoordinates)) {
+            onDragStart(calculateOffset(componentCoordinates, windowDragCoordinates!!))
+        }
+        return idsCounter++
+    }
+
+    /**
+     * Unsubscribes handler with [handleId].
+     * Calls [ComponentDragHandler.onDragCancel] if drag is going and handler's component is under pointer
+     *
+     * Disable drag handling for [window] if there are no more handlers.
+     *
+     * @param handleId id provided by [installComponentDragHandler] function
+     */
+    fun stopDragHandling(handleId: Int) {
+        val handler = handlers[handleId]
+        if (handler != null && isExternalDragInsideComponent(handler.componentCoordinates, windowDragCoordinates)) {
+            handler.onDragCancel()
+        }
+        handlers.remove(handleId)
+
+        if (handlers.isEmpty()) {
+            isActive = false
+        }
+    }
+
+
+    private class ComponentDragHandler(
+        val componentCoordinates: LayoutCoordinates,
+        val onDragStart: (Offset) -> Unit,
+        val onDrag: (Offset) -> Unit,
+        val onDragCancel: () -> Unit,
+        val onDrop: (DropData) -> Unit
+    )
+
+    companion object {
+        private fun isExternalDragInsideComponent(
+            componentCoordinates: LayoutCoordinates?,
+            windowDragCoordinates: Offset?
+        ): Boolean {
+            if (componentCoordinates == null || windowDragCoordinates == null) {
+                return false
+            }
+
+            return componentCoordinates.boundsInWindow().contains(windowDragCoordinates)
+        }
+
+        private fun calculateOffset(
+            componentCoordinates: LayoutCoordinates,
+            windowDragCoordinates: Offset
+        ): Offset {
+            return componentCoordinates.windowToLocal(windowDragCoordinates)
+        }
+    }
+}
+
+// @VisibleForTesting
+@OptIn(ExperimentalComposeUiApi::class)
+internal class AwtWindowDragTargetListener(
+    private val window: Window,
+    val onDragEnterWindow: (Offset) -> Unit,
+    val onDragInsideWindow: (Offset) -> Unit,
+    val onDragExit: () -> Unit,
+    val onDrop: (DropData) -> Boolean,
+) : DropTargetListener {
+    private val density = window.density.density
+
+    override fun dragEnter(dtde: DropTargetDragEvent) {
+        onDragEnterWindow(dtde.location.windowOffset())
+    }
+
+    override fun dragOver(dtde: DropTargetDragEvent) {
+        onDragInsideWindow(dtde.location.windowOffset())
+    }
+
+    // takes title bar and other insets into account
+    private fun Point.windowOffset(): Offset {
+        val offsetX = (x - window.insets.left) * density
```

Comment:
I don't know whether it matters, but `window.density.density` is `window.graphicsConfiguration.defaultTransform.scaleX`. Maybe you'd want to use `scaleY` for the top insets.

Reply:
This is a good point! Firstly I used scaleY here, but than I looked at other places and saw that window.density.density is used everywhere, so I think we should be consistent and if user uses different scales for X and Y (which is weird?) we have to fix it everywhere at once. What do you think?

Reply:
We usually don't distinguish between scaleX/scaleY, as there is no use cases for that in Compose. We retrieve the screen density from this property because there is no other way.

Reply:
So, I'll keep it, right?